{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ebea1c0d-c0ee-4bb9-a9cc-578929a696c7",
   "metadata": {},
   "source": [
    "## Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "\n",
    "Exception class is the base class for exception and all the exception must be derived from base exception class. Base exception\n",
    "class has built in exception which can be easily extended into child class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab2e31bc-b826-44c5-8e5b-dda4fddbd8b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The class hierarchy for built-in exceptions is:\n",
      " Exception\n",
      "... TypeError\n",
      "...... FloatOperation\n",
      "...... MultipartConversionError\n",
      "... StopAsyncIteration\n",
      "... StopIteration\n",
      "... ImportError\n",
      "...... ModuleNotFoundError\n",
      "...... ZipImportError\n",
      "... OSError\n",
      "...... ConnectionError\n",
      "......... BrokenPipeError\n",
      "......... ConnectionAbortedError\n",
      "......... ConnectionRefusedError\n",
      "......... ConnectionResetError\n",
      "............ RemoteDisconnected\n",
      "...... BlockingIOError\n",
      "...... ChildProcessError\n",
      "...... FileExistsError\n",
      "...... FileNotFoundError\n",
      "...... IsADirectoryError\n",
      "...... NotADirectoryError\n",
      "...... InterruptedError\n",
      "......... InterruptedSystemCall\n",
      "...... PermissionError\n",
      "...... ProcessLookupError\n",
      "...... TimeoutError\n",
      "...... UnsupportedOperation\n",
      "...... itimer_error\n",
      "...... herror\n",
      "...... gaierror\n",
      "...... SSLError\n",
      "......... SSLCertVerificationError\n",
      "......... SSLZeroReturnError\n",
      "......... SSLWantWriteError\n",
      "......... SSLWantReadError\n",
      "......... SSLSyscallError\n",
      "......... SSLEOFError\n",
      "...... Error\n",
      "......... SameFileError\n",
      "...... SpecialFileError\n",
      "...... ExecError\n",
      "...... ReadError\n",
      "...... URLError\n",
      "......... HTTPError\n",
      "......... ContentTooShortError\n",
      "...... BadGzipFile\n",
      "... EOFError\n",
      "...... IncompleteReadError\n",
      "... RuntimeError\n",
      "...... RecursionError\n",
      "...... NotImplementedError\n",
      "......... ZMQVersionError\n",
      "......... StdinNotImplementedError\n",
      "...... _DeadlockError\n",
      "...... BrokenBarrierError\n",
      "...... BrokenExecutor\n",
      "......... BrokenThreadPool\n",
      "...... SendfileNotAvailableError\n",
      "...... ExtractionError\n",
      "...... VariableError\n",
      "... NameError\n",
      "...... UnboundLocalError\n",
      "... AttributeError\n",
      "...... FrozenInstanceError\n",
      "... SyntaxError\n",
      "...... IndentationError\n",
      "......... TabError\n",
      "... LookupError\n",
      "...... IndexError\n",
      "...... KeyError\n",
      "......... NoSuchKernel\n",
      "......... UnknownBackend\n",
      "...... CodecRegistryError\n",
      "... ValueError\n",
      "...... UnicodeError\n",
      "......... UnicodeEncodeError\n",
      "......... UnicodeDecodeError\n",
      "......... UnicodeTranslateError\n",
      "...... UnsupportedOperation\n",
      "...... JSONDecodeError\n",
      "...... SSLCertVerificationError\n",
      "...... Error\n",
      "...... UnsupportedDigestmodError\n",
      "...... IllegalMonthError\n",
      "...... IllegalWeekdayError\n",
      "...... ParserError\n",
      "...... ClassNotFound\n",
      "...... ClipboardEmpty\n",
      "...... MessageDefect\n",
      "......... NoBoundaryInMultipartDefect\n",
      "......... StartBoundaryNotFoundDefect\n",
      "......... CloseBoundaryNotFoundDefect\n",
      "......... FirstHeaderLineIsContinuationDefect\n",
      "......... MisplacedEnvelopeHeaderDefect\n",
      "......... MissingHeaderBodySeparatorDefect\n",
      "......... MultipartInvariantViolationDefect\n",
      "......... InvalidMultipartContentTransferEncodingDefect\n",
      "......... UndecodableBytesDefect\n",
      "......... InvalidBase64PaddingDefect\n",
      "......... InvalidBase64CharactersDefect\n",
      "......... InvalidBase64LengthDefect\n",
      "......... HeaderDefect\n",
      "............ InvalidHeaderDefect\n",
      "............ HeaderMissingRequiredValue\n",
      "............ NonPrintableDefect\n",
      "............ ObsoleteHeaderDefect\n",
      "............ NonASCIILocalPartDefect\n",
      "............ InvalidDateDefect\n",
      "...... MacroToEdit\n",
      "...... InvalidFileException\n",
      "...... UnequalIterablesError\n",
      "...... InvalidVersion\n",
      "...... _InvalidELFFileHeader\n",
      "...... InvalidWheelFilename\n",
      "...... InvalidSdistFilename\n",
      "...... InvalidSpecifier\n",
      "...... InvalidMarker\n",
      "...... UndefinedComparison\n",
      "...... UndefinedEnvironmentName\n",
      "...... InvalidRequirement\n",
      "......... RequirementParseError\n",
      "...... InvalidVersion\n",
      "... AssertionError\n",
      "... ArithmeticError\n",
      "...... FloatingPointError\n",
      "...... OverflowError\n",
      "...... ZeroDivisionError\n",
      "......... DivisionByZero\n",
      "......... DivisionUndefined\n",
      "...... DecimalException\n",
      "......... Clamped\n",
      "......... Rounded\n",
      "............ Underflow\n",
      "............ Overflow\n",
      "......... Inexact\n",
      "............ Underflow\n",
      "............ Overflow\n",
      "......... Subnormal\n",
      "............ Underflow\n",
      "......... DivisionByZero\n",
      "......... FloatOperation\n",
      "......... InvalidOperation\n",
      "............ ConversionSyntax\n",
      "............ DivisionImpossible\n",
      "............ DivisionUndefined\n",
      "............ InvalidContext\n",
      "... SystemError\n",
      "...... CodecRegistryError\n",
      "... ReferenceError\n",
      "... MemoryError\n",
      "... BufferError\n",
      "... Warning\n",
      "...... UserWarning\n",
      "......... GetPassWarning\n",
      "......... FormatterWarning\n",
      "...... EncodingWarning\n",
      "...... DeprecationWarning\n",
      "......... ProvisionalWarning\n",
      "...... PendingDeprecationWarning\n",
      "...... SyntaxWarning\n",
      "...... RuntimeWarning\n",
      "......... ProactorSelectorThreadWarning\n",
      "......... UnknownTimezoneWarning\n",
      "......... PEP440Warning\n",
      "...... FutureWarning\n",
      "......... ProvisionalCompleterWarning\n",
      "...... ImportWarning\n",
      "...... UnicodeWarning\n",
      "...... BytesWarning\n",
      "...... ResourceWarning\n",
      "...... DeprecatedTzFormatWarning\n",
      "...... PkgResourcesDeprecationWarning\n",
      "... _OptionError\n",
      "... _Error\n",
      "... error\n",
      "... Verbose\n",
      "... Error\n",
      "... SubprocessError\n",
      "...... CalledProcessError\n",
      "...... TimeoutExpired\n",
      "... TokenError\n",
      "... StopTokenizing\n",
      "... ClassFoundException\n",
      "... EndOfBlock\n",
      "... TraitError\n",
      "... Error\n",
      "... Error\n",
      "...... CancelledError\n",
      "...... TimeoutError\n",
      "...... InvalidStateError\n",
      "... _GiveupOnSendfile\n",
      "... error\n",
      "... Incomplete\n",
      "... TimeoutError\n",
      "... InvalidStateError\n",
      "... LimitOverrunError\n",
      "... QueueEmpty\n",
      "... QueueFull\n",
      "... Empty\n",
      "... Full\n",
      "... ArgumentError\n",
      "... ZMQBaseError\n",
      "...... ZMQError\n",
      "......... ContextTerminated\n",
      "......... Again\n",
      "......... InterruptedSystemCall\n",
      "...... ZMQBindError\n",
      "...... NotDone\n",
      "... PickleError\n",
      "...... PicklingError\n",
      "...... UnpicklingError\n",
      "... _Stop\n",
      "... ArgumentError\n",
      "... ArgumentTypeError\n",
      "... ConfigError\n",
      "...... ConfigLoaderError\n",
      "......... ArgumentError\n",
      "...... ConfigFileNotFound\n",
      "... ConfigurableError\n",
      "...... MultipleInstanceError\n",
      "... ApplicationError\n",
      "... error\n",
      "... TimeoutError\n",
      "... error\n",
      "... ReturnValueIgnoredError\n",
      "... KeyReuseError\n",
      "... UnknownKeyError\n",
      "... LeakedCallbackError\n",
      "... BadYieldError\n",
      "... ReturnValueIgnoredError\n",
      "... Return\n",
      "... InvalidPortNumber\n",
      "... error\n",
      "... LZMAError\n",
      "... RegistryError\n",
      "... _GiveupOnFastCopy\n",
      "... Error\n",
      "...... NoSectionError\n",
      "...... DuplicateSectionError\n",
      "...... DuplicateOptionError\n",
      "...... NoOptionError\n",
      "...... InterpolationError\n",
      "......... InterpolationMissingOptionError\n",
      "......... InterpolationSyntaxError\n",
      "......... InterpolationDepthError\n",
      "...... ParsingError\n",
      "......... MissingSectionHeaderError\n",
      "... NoIPAddresses\n",
      "... BadZipFile\n",
      "... LargeZipFile\n",
      "... BadEntryPoint\n",
      "... NoSuchEntryPoint\n",
      "... DuplicateKernelError\n",
      "... ErrorDuringImport\n",
      "... NotOneValueFound\n",
      "... CannotEval\n",
      "... OptionError\n",
      "... BdbQuit\n",
      "... Restart\n",
      "... ExceptionPexpect\n",
      "...... EOF\n",
      "...... TIMEOUT\n",
      "... PtyProcessError\n",
      "... FindCmdError\n",
      "... HomeDirError\n",
      "... ProfileDirError\n",
      "... IPythonCoreError\n",
      "...... TryNext\n",
      "...... UsageError\n",
      "...... StdinNotImplementedError\n",
      "... InputRejected\n",
      "... GetoptError\n",
      "... ErrorToken\n",
      "... PrefilterError\n",
      "... AliasError\n",
      "...... InvalidAliasError\n",
      "... Error\n",
      "...... InterfaceError\n",
      "...... DatabaseError\n",
      "......... InternalError\n",
      "......... OperationalError\n",
      "......... ProgrammingError\n",
      "......... IntegrityError\n",
      "......... DataError\n",
      "......... NotSupportedError\n",
      "... Warning\n",
      "... SpaceInInput\n",
      "... DOMException\n",
      "...... IndexSizeErr\n",
      "...... DomstringSizeErr\n",
      "...... HierarchyRequestErr\n",
      "...... WrongDocumentErr\n",
      "...... InvalidCharacterErr\n",
      "...... NoDataAllowedErr\n",
      "...... NoModificationAllowedErr\n",
      "...... NotFoundErr\n",
      "...... NotSupportedErr\n",
      "...... InuseAttributeErr\n",
      "...... InvalidStateErr\n",
      "...... SyntaxErr\n",
      "...... InvalidModificationErr\n",
      "...... NamespaceErr\n",
      "...... InvalidAccessErr\n",
      "...... ValidationErr\n",
      "... ValidationError\n",
      "... EditReadOnlyBuffer\n",
      "... _Retry\n",
      "... InvalidLayoutError\n",
      "... HeightIsUnknownError\n",
      "... ParserSyntaxError\n",
      "... InternalParseError\n",
      "... _PositionUpdatingFinished\n",
      "... SimpleGetItemNotFound\n",
      "... UncaughtAttributeError\n",
      "... HasNoContext\n",
      "... ParamIssue\n",
      "... _JediError\n",
      "...... InternalError\n",
      "...... WrongVersion\n",
      "...... RefactoringError\n",
      "... OnErrorLeaf\n",
      "... InvalidPythonEnvironment\n",
      "... MessageError\n",
      "...... MessageParseError\n",
      "......... HeaderParseError\n",
      "......... BoundaryError\n",
      "...... MultipartConversionError\n",
      "...... CharsetError\n",
      "... Error\n",
      "... HTTPException\n",
      "...... NotConnected\n",
      "...... InvalidURL\n",
      "...... UnknownProtocol\n",
      "...... UnknownTransferEncoding\n",
      "...... UnimplementedFileMode\n",
      "...... IncompleteRead\n",
      "...... ImproperConnectionState\n",
      "......... CannotSendRequest\n",
      "......... CannotSendHeader\n",
      "......... ResponseNotReady\n",
      "...... BadStatusLine\n",
      "......... RemoteDisconnected\n",
      "...... LineTooLong\n",
      "... InteractivelyDefined\n",
      "... KillEmbedded\n",
      "... Error\n",
      "...... NoSuchProcess\n",
      "......... ZombieProcess\n",
      "...... AccessDenied\n",
      "...... TimeoutExpired\n",
      "... _Ipv6UnsupportedError\n",
      "... QueueEmpty\n",
      "... QueueFull\n",
      "... DebuggerInitializationError\n",
      "... ExpatError\n",
      "... Error\n",
      "...... ProtocolError\n",
      "...... ResponseError\n",
      "...... Fault\n",
      "... ParseBaseException\n",
      "...... ParseException\n",
      "...... ParseFatalException\n",
      "......... ParseSyntaxException\n",
      "... RecursiveGrammarException\n",
      "... ResolutionError\n",
      "...... VersionConflict\n",
      "......... ContextualVersionConflict\n",
      "...... DistributionNotFound\n",
      "...... UnknownExtra\n",
      "... _Error\n",
      "... UnableToResolveVariableException\n",
      "... InvalidTypeInArgsException\n"
     ]
    }
   ],
   "source": [
    "\n",
    "## Q2. Write a python program to print Python Exception Hierarchy\n",
    "\n",
    "\n",
    "import inspect\n",
    "print(\"The class hierarchy for built-in exceptions is:\")\n",
    "inspect.getclasstree(inspect.getmro(Exception))\n",
    "def classtree(cls, indent=0):\n",
    "    print('.' * indent, cls.__name__)\n",
    "    for subcls in cls.__subclasses__():\n",
    "        classtree(subcls, indent + 3)\n",
    "classtree(Exception)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "19382cf8-bfa6-4e09-8c9a-5ed408513c2f",
   "metadata": {},
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 11\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m## Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m## ArithmeticError class includes errors which are specific to  numeric operations. Error includes zerodivision error, overflow error,floating point error etc.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m      6\u001b[0m \n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# 1) Zerodivision Error : Ocuurs when we divide a number by 0.\u001b[39;00m\n\u001b[1;32m      9\u001b[0m i \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m5\u001b[39m\n\u001b[0;32m---> 11\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mi\u001b[49m\u001b[38;5;241;43m/\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m)\n",
      "\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "## Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "\n",
    "## ArithmeticError class includes errors which are specific to  numeric operations. Error includes zerodivision error, overflow error,floating point error etc.\n",
    "\n",
    "\n",
    "\n",
    "# 1) Zerodivision Error : Ocuurs when we divide a number by 0.\n",
    "\n",
    "i = 5\n",
    "\n",
    "print(i/0)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ece4d9d6-a053-4469-aa29-f65653d412ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.0\n",
      "25.0\n",
      "15625.0\n",
      "5.960464477539062e+16\n",
      "7.52316384526264e+83\n"
     ]
    },
    {
     "ename": "OverflowError",
     "evalue": "(34, 'Numerical result out of range')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOverflowError\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m j \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m5.0\u001b[39m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m1000\u001b[39m):\n\u001b[0;32m----> 8\u001b[0m     j \u001b[38;5;241m=\u001b[39m \u001b[43mj\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mi\u001b[49m\n\u001b[1;32m      9\u001b[0m     \u001b[38;5;28mprint\u001b[39m(j)\n",
      "\u001b[0;31mOverflowError\u001b[0m: (34, 'Numerical result out of range')"
     ]
    }
   ],
   "source": [
    "# 2) OverflowError exception \n",
    "## An OverflowError exception is raised when an arithmetic operation exceeds the limits to be represented. \n",
    "## in Below example, overflow exception occured as result is out of range\n",
    "\n",
    "j = 5.0\n",
    "\n",
    "for i in range(1, 1000):\n",
    "    j = j**i\n",
    "    print(j)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "38b6330b-4729-40d8-a815-f15629188b0e",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 10\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m## Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m## The LookupError exception in Python forms the base class for all exceptions that are raised \u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m## when an index or a key is not found for a sequence or dictionary respectively.\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m## Example\u001b[39;00m\n\u001b[1;32m      8\u001b[0m list1 \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m1\u001b[39m,\u001b[38;5;241m2\u001b[39m,\u001b[38;5;241m3\u001b[39m,\u001b[38;5;241m4\u001b[39m]\n\u001b[0;32m---> 10\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mlist1\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m]\u001b[49m)\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "## Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "\n",
    "## The LookupError exception in Python forms the base class for all exceptions that are raised \n",
    "## when an index or a key is not found for a sequence or dictionary respectively.\n",
    "\n",
    "## Example\n",
    "\n",
    "## IndexError occurs when the index we are checking doesnot exist in the list\n",
    "\n",
    "list1 = [1,2,3,4]\n",
    "\n",
    "print(list1[5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "39394b56-fa65-48fd-b729-beca5f20fcaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sambav\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'weight'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m person \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m\"\u001b[39m:\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msambav\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mage\u001b[39m\u001b[38;5;124m\"\u001b[39m:\u001b[38;5;241m30\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlanguage\u001b[39m\u001b[38;5;124m\"\u001b[39m:\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPython\u001b[39m\u001b[38;5;124m\"\u001b[39m}\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28mprint\u001b[39m(person[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m----> 6\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mperson\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mweight\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'weight'"
     ]
    }
   ],
   "source": [
    "## KeyError occurs when the key which is being searched is not exist in the dictnoary\n",
    "\n",
    "person = {\"name\":\"sambav\",\"age\":30,\"language\":\"Python\"}\n",
    "\n",
    "print(person[\"name\"])\n",
    "print(person[\"weight\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "82937741-76f0-454a-a443-f158f7c30515",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'sambav'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[22], line 13\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m## Q5. Explain ImportError. What is ModuleNotFoundError?\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m## This error generally occurs when a class cannot be imported due to one of the following reasons:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m      9\u001b[0m \n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m## ModulenotFound error occurs when a module doesnot exist in python library\u001b[39;00m\n\u001b[0;32m---> 13\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01msambav\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'sambav'"
     ]
    }
   ],
   "source": [
    "## Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "\n",
    "## This error generally occurs when a class cannot be imported due to one of the following reasons:\n",
    "### The imported class is in a circular dependency.\n",
    "### The imported class is unavailable or was not created.\n",
    "### The imported class name is misspelled.\n",
    "### The imported class from a module is misplaced.\n",
    "### The imported class is unavailable in the Python library\n",
    "\n",
    "## ModulenotFound error occurs when a module doesnot exist in python library\n",
    "\n",
    "\n",
    "import sambav"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "49fb4b9e-b33e-4827-bc1b-8d460ffb73ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Q6. List down some best practices for exception handling in python.\n",
    "\n",
    "## 1) Never use irrelavant and unneccesary exception handling.\n",
    "## 2) use always a specific exception\n",
    "## 3) Print/log always with a valid message.\n",
    "## 4) Clean up all the resource at the end of the program\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a37f1811-4695-48ae-843e-e511a097985d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
